src/smp.rs

```diff
@@ -54,15 +62,45 @@ impl PerCpu {
         VmIdTracker::init();
 
         // Now initialize each PerCpu structure.
+        // Unwrap okay: PER_CPU_PAGES is non zero.
+        let mut pcpu_stacks =
+            pcpu_pages.into_chunks_iter(core::num::NonZeroU64::new(PER_CPU_PAGES).unwrap());
         for i in 0..cpu_info.num_cpus() {
             let cpu_id = CpuId::new(i);
+            // Unwrap okay. We allocated the area `with num_cpus() * PER_CPU_PAGES` pages.
+            let pcpu_stack = pcpu_stacks.next().unwrap();
+            // Boot CPU is special. Doesn't use the PCPU_BASE area as stack, only for the PCPU area.
+            // TODO: Do not allocate PCPU area for boot cpu.
+            let stack_pages = if cpu_id == boot_cpu {
+                Self::boot_cpu_stack()
```

Comment:
i think we discussed this offline and i'm fuzzy on the details, but why can't we switch tthe boot CPU over to use a stack from `pcpu_stacks`? the whole boot cpu having a different stack size than the secondary cpus thing has hidden a couple of bugs from us so far. (it would also reduce all the special-casing you have to do here.)

Reply:
Sure, the problem with this is that we would still need a (huge) stack at the beginning (init phase). That stack is allocated by the linker.

I implemented both ways, one where the bootstrap CPU used this PerCPu stack (which simplifies things quite a lot on a architectural point of view) but that wastes the original init-time stack memory.

With this, we can have a following PR that saves stack memory by changing this `smp` code to not allocate a full blown CPU stack just for the PerCpu data stucture. There are various ways to achieve this, non particularly difficult.

Reply:
Now PerCpu data and stack are separate.

Reply:
understood re: not wanting to waste memory. but should we at least clamp the size of the boot CPU's stack when we set up the VA mapping so that all CPUs have the same stack size? i really want to get rid of this situation where things sometimes work because they're run on the boot CPU vs secondary CPU.

Reply:
I do that in HypMap on stack region. I return InvalidStackSize if the linker allocated stack is not _exactly_ `HYP_STACK_PAGES`. But thinking about it, possibly clamping it afterwards could be more generic and allows us a bigger boot stack than runtime requires. Would be wasted memory though.

Although, I kind of like to remove any power to the linker and check at runtime if it was built properly.